include/quic/parser.hpp

```diff
@@ -0,0 +1,378 @@
+            log::debug(bp_cat, "{} called", __PRETTY_FUNCTION__);
+            sent_reqs.clear();
+            s.close(io_error{ec});
+        };
+
+      public:
+        template <typename... Opt>
+        explicit bparser(Connection& _c, Endpoint& _e, Opt&&... opts) : Stream{_c, _e}
+        {
+            ((void)handle_bp_opt(std::forward<Opt>(opts)), ...);
+        }
+
+        void request(std::string endpoint, std::string body) override
+        {
+            log::debug(bp_cat, "{} called", __PRETTY_FUNCTION__);
+
+            auto req = make_request(std::move(endpoint), std::move(body));
+            send(req->view());
+
+            auto& sr = sent_reqs[req->timeout];
+            sr = std::move(req);
+        }
+
+        void command(std::string endpoint, std::string body) override
+        {
+            log::debug(bp_cat, "{} called", __PRETTY_FUNCTION__);
+
+            auto req = make_command(std::move(endpoint), std::move(body));
+            send(req->payload());
+        }
+
+        void respond(std::string rid, std::string body, bool error = false) override
+        {
+            log::debug(bp_cat, "{} called", __PRETTY_FUNCTION__);
+
+            auto req = make_response(std::move(rid), std::move(body), error);
+            send(req->payload());
+        }
+
+        void receive(bstring_view data) override
+        {
+            log::info(bp_cat, "bparser recv data callback called!");
+            log::debug(bp_cat, "Received data: {}", buffer_printer{data});
+
+            try
+            {
+                process_incoming(to_sv(data));
+            }
+            catch (std::exception& e)
+            {
+                log::error(bp_cat, "Exception caught: {}", e.what());
+            }
+        }
+
+        void closed(uint64_t app_code) override
+        {
+            log::info(bp_cat, "bparser close callback called!");
+            close_callback(*this, app_code);
+        }
+
+        void check_timeouts() override
+        {
+            const auto& now = get_time();
+
+            for (auto itr = sent_reqs.begin(); itr != sent_reqs.end();)
+            {
+                if (itr->first < now)
+                    itr = sent_reqs.erase(itr);
+                else
+                    return;
+            }
+        }
+
+      private:
+        void handle_bp_opt(std::function<void(Stream&, message)> recv_cb)
+        {
+            log::debug(bp_cat, "Bparser set user-provided recv callback!");
+            recv_callback = std::move(recv_cb);
+        }
+
+        void handle_bp_opt(std::function<void(Stream&, uint64_t)> close_cb)
+        {
+            log::debug(bp_cat, "Bparser set user-provided close callback!");
+            close_callback = std::move(close_cb);
+        }
+
+        bool match(std::string_view rid)
+        {
+            log::trace(bp_cat, "{} called", __PRETTY_FUNCTION__);
+
+            for (auto& r : sent_reqs)
```

Comment:
This linear scan is a little unfortunately (e.g. think SPNS where we have thousands of requests out and arriving at once).

Thinking about what we could do instead...

Reply:
Timestamps and request id's are both monotonically increasing, so we could binary search for log(n) complexity

Reply:
Currently is an ordered map of `request time -> request` which is good for clearing timed-out requests, but means matching request id is slower.  Could instead map `request id -> request`.  If `request id` is monotonically increasing, time should as well, so the same clear timeout logic should work (except where request id would wrap, but...64-bit integer.)

Reply:
Yeah.  Thinking about that it makes sense: we probably will have to do something along the lines of (pseudocode):

while first entry is expired:
    trigger failure callback for first entry
    pop first entry

and so for expiring, all we really care about is that they are ordered, but for lookup we're operating on the id, so id as key seems to make sense to me.

Or hell, maybe we could just go back to std::deque and do a std::binary_search that compares the `id` on lookup, and just pop off the front (as pseudocode above) when we look for expired stuff.